utils/merge_syft_sbom.py

```diff
@@ -0,0 +1,128 @@
+#!/usr/bin/env python3
+import json
+import sys
+
+from typing import Any, Callable
+from urllib.parse import urlsplit, quote_plus
+
+
+# Functions to detect components that are reported differently by Syft
+def _is_syft_local_golang_component(component: dict) -> bool:
+    """
+    Check if a Syft Golang reported component is a local replacement.
+
+    Local replacements are reported in a very different way by Cachi2,
+    which is why the same reports by Syft should be removed.
+    """
+    return (
+        component.get("purl", "").startswith("pkg:golang")
+        and component.get("name", "").startswith(".")
+        or component.get("version", "") == "(devel)"
+    )
+
+
+def _is_cachi2_non_pypi_pip_dependency(component: dict) -> bool:
+    """
+    Check if Cachi2 Pip component was fetched a VCS or HTTPS location.
+
+    Cachi2 reports non-pypi components in a different way from Syft, so
+    the reports from Syft need to be removed.
+
+    Unfortunately, there's no way to determine which components are non-Pypi
+    by looking at the Syft report alone. This function is meant to create a
+    list of non-Pypi components from Cachi2's SBOM, then remove the ones reported
+    by Syft in the merged SBOM.
+    """
+    return (
+        component.get("purl", "").startswith("pkg:pypi")
+        and "version" not in component
+    )
```

Comment:
Maybe checking for vcs_url | download_url qualifiers would be more robust and re-usable for npm?

Reply:
Good point. I did not do any research on how NPM dedup would look like, though. I'll look into this tomorrow, and I bet it add more dedup rules to this already long script :confused: 

We can revisit this topic then.

Reply:
Ok, so NPM rule is added.

Troblem about defaulting to `vcs_url | download_url` is that we only check for Cachi2 non-registry components for Pip. For NPM, we check for Syft non-registry components, and for Golang, we only care about local replacements.

Reply:
Updating this thread here: we have decided to use the same rule to dedup non-registry components for both NPM and Pip (that is, check Cachi2 reported non-registry components, and remove all Syft components that have a matching name). So, yeah, it makes sense to use `vcs_url | download_url` qualifiers.